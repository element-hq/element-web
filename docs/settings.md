# Settings Reference

This document serves as developer documentation for using "Granular Settings". Granular Settings allow users to specify
different values for a setting at particular levels of interest. For example, a user may say that in a particular room
they want URL previews off, but in all other rooms they want them enabled. The `SettingsStore` helps mask the complexity
of dealing with the different levels and exposes easy to use getters and setters.

## Levels

Granular Settings rely on a series of known levels in order to use the correct value for the scenario. These levels, in
order of priority, are:

-   `device` - The current user's device
-   `room-device` - The current user's device, but only when in a specific room
-   `room-account` - The current user's account, but only when in a specific room
-   `account` - The current user's account
-   `room` - A specific room (setting for all members of the room)
-   `config` - Values are defined by the `setting_defaults` key (usually) in `config.json`
-   `default` - The hardcoded default for the settings

Individual settings may control which levels are appropriate for them as part of the defaults. This is often to ensure
that room administrators cannot force account-only settings upon participants.

## Settings

Settings are the different options a user may set or experience in the application. These are pre-defined in
`src/settings/Settings.tsx` under the `SETTINGS` constant, and match the `ISetting` interface as defined there.

Settings that support the config level can be set in the config file under the `setting_defaults` key (note that some
settings, like the "theme" setting, are special cased in the config file):

```json5
{
  ...
  "setting_defaults": {
    "settingName": true
  },
  ...
}
```

### Getting values for a setting

After importing `SettingsStore`, simply make a call to `SettingsStore.getValue`. The `roomId` parameter should always
be supplied where possible, even if the setting does not have a per-room level value. This is to ensure that the value
returned is best represented in the room, particularly if the setting ever gets a per-room level in the future.

In settings pages it is often desired to have the value at a particular level instead of getting the calculated value.
Call `SettingsStore.getValueAt` to get the value of a setting at a particular level, and optionally make it explicitly
at that level. By default `getValueAt` will traverse the tree starting at the provided level; making it explicit means
it will not go beyond the provided level. When using `getValueAt`, please be sure to use `SettingLevel` to represent the
target level.

### Setting values for a setting

Values are defined at particular levels and should be done in a safe manner. There are two checks to perform to ensure a
clean save: is the level supported and can the user actually set the value. In most cases, neither should be an issue
although there are circumstances where this changes. An example of a safe call is:

```javascript
const isSupported = SettingsStore.isLevelSupported(SettingLevel.ROOM);
if (isSupported) {
    const canSetValue = SettingsStore.canSetValue("mySetting", "!curbf:matrix.org", SettingLevel.ROOM);
    if (canSetValue) {
        SettingsStore.setValue("mySetting", "!curbf:matrix.org", SettingLevel.ROOM, newValue);
    }
}
```

These checks may also be performed in different areas of the application to avoid the verbose example above. For
instance, the component which allows changing the setting may be hidden conditionally on the above conditions.

##### `SettingsFlag` component

Where possible, the `SettingsFlag` component should be used to set simple "flip-a-bit" (true/false) settings. The
`SettingsFlag` also supports simple radio button options, such as the theme the user would like to use.

```TSX
<SettingsFlag name="theSettingId" level={SettingsLevel.ROOM} roomId="!curbf:matrix.org"
    label={_td("Your label here")} // optional, if falsey then the `SettingsStore` will be used
    onChange={function(newValue) { }} // optional, called after saving
    isExplicit={false} // this is passed along to `SettingsStore.getValueAt`, defaulting to false
    manualSave={false} // if true, saving is delayed. You will need to call .save() on this component

    // Options for radio buttons
    group="your-radio-group" // this enables radio button support
    value="yourValueHere" // the value for this particular option
/>
```

### Getting the display name for a setting

Simply call `SettingsStore.getDisplayName`. The appropriate display name will be returned and automatically translated
for you. If a display name cannot be found, it will return `null`.

## Features

Feature flags are just like regular settings with some underlying semantics for how they are meant to be used. Usually
a feature flag is used when a portion of the application is under development or not ready for full release yet, such
as new functionality or experimental ideas. In these cases, the feature name _should_ be named with the `feature_*`
convention and must be tagged with `isFeature: true` in the setting definition. By doing so, the feature will automatically
appear in the "labs" section of the user's settings.

Features can be controlled at the config level using the following structure:

```json
"features": {
  "feature_lazyloading": true
}
```

When `true`, the user will see the feature as enabled. Similarly, when `false` the user will see the feature as disabled.
The user will only be able to change/see these states if `show_labs_settings: true` is in the config.

### Determining if a feature is enabled

Call `SettingsStore.getValue()` as you would for any other setting.

### Enabling a feature

Call `SettingsStore.setValue("feature_name", null, SettingLevel.DEVICE, true)`.

### A note on UI features

UI features are a different concept to plain features. Instead of being representative of unstable or
unpredicatable behaviour, they are logical chunks of UI which can be disabled by deployments for ease
of understanding with users. They are simply represented as boring settings with a convention of being
named as `UIFeature.$location` where `$location` is a rough descriptor of what is toggled, such as
`URLPreviews` or `Communities`.

UI features also tend to have their own setting controller (see below) to manipulate settings which might
be affected by the UI feature being disabled. For example, if URL previews are disabled as a UI feature
then the URL preview options will use the `UIFeatureController` to ensure they remain disabled while the
UI feature is disabled.

## Setting controllers

Settings may have environmental factors that affect their value or need additional code to be called when they are
modified. A setting controller is able to override the calculated value for a setting and react to changes in that
setting. Controllers are not a replacement for the level handlers and should only be used to ensure the environment is
kept up to date with the setting where it is otherwise not possible. An example of this is the notification settings:
they can only be considered enabled if the platform supports notifications, and enabling notifications requires
additional steps to actually enable notifications.

For more information, see `src/settings/controllers/SettingController.ts`.

## Local echo

`SettingsStore` will perform local echo on all settings to ensure that immediately getting values does not cause a
split-brain scenario. As mentioned in the "Setting values for a setting" section, the appropriate checks should be done
to ensure that the user is allowed to set the value. The local echo system assumes that the user has permission and that
the request will go through successfully. The local echo only takes effect until the request to save a setting has
completed (either successfully or otherwise).

```javascript
SettingsStore.setValue(...).then(() => {
  // The value has actually been stored at this point.
});
SettingsStore.getValue(...); // this will return the value set in `setValue` above.
```

## Watching for changes

Most use cases do not need to set up a watcher because they are able to react to changes as they are made, or the
changes which are made are not significant enough for it to matter. Watchers are intended to be used in scenarios where
it is important to react to changes made by other logged in devices. Typically, this would be done within the component
itself, however the component should not be aware of the intricacies of setting inversion or remapping to particular
data structures. Instead, a generic watcher interface is provided on `SettingsStore` to watch (and subsequently unwatch)
for changes in a setting.

An example of a watcher in action would be:

```javascript
class MyComponent extends React.Component {
    settingWatcherRef = null;

    componentDidMount() {
        const callback = (settingName, roomId, level, newValAtLevel, newVal) => {
            this.setState({ color: newVal });
        };
        this.settingWatcherRef = SettingsStore.watchSetting("roomColor", "!example:matrix.org", callback);
    }

    componentWillUnmount() {
        SettingsStore.unwatchSetting(this.settingWatcherRef);
    }
}
```

# Maintainers Reference

The granular settings system has a few complex parts to power it. This section is to document how the `SettingsStore` is
supposed to work.

### General information

The `SettingsStore` uses the hardcoded `LEVEL_ORDER` constant to ensure that it is using the correct override procedure.
The array is checked from left to right, simulating the behaviour of overriding values from the higher levels. Each
level should be defined in this array, including `default`.

Handlers (`src/settings/handlers/SettingsHandler.ts`) represent a single level and are responsible for getting and
setting values at that level. Handlers also provide additional information to the `SettingsStore` such as if the level
is supported or if the current user may set values at the level. The `SettingsStore` will use the handler to enforce
checks and manipulate settings. Handlers are also responsible for dealing with migration patterns or legacy settings for
their level (for example, a setting being renamed or using a different key from other settings in the underlying store).
Handlers are provided to the `SettingsStore` via the `LEVEL_HANDLERS` constant. `SettingsStore` will optimize lookups by
only considering handlers that are supported on the platform.

Local echo is achieved through `src/settings/handlers/LocalEchoWrapper.ts` which acts as a wrapper around a given
handler. This is automatically applied to all defined `LEVEL_HANDLERS` and proxies the calls to the wrapped handler
where possible. The echo is achieved by a simple object cache stored within the class itself. The cache is invalidated
immediately upon the proxied save call succeeding or failing.

Controllers are notified of changes by the `SettingsStore`, and are given the opportunity to override values after the
`SettingsStore` has deemed the value calculated. Controllers are invoked as the last possible step in the code.

### Features

See above for feature reference.

### Watchers

Watchers can appear complicated under the hood: there is a central `WatchManager` which handles the actual invocation
of callbacks, and callbacks are managed by the SettingsStore by redirecting the caller's callback to a dedicated
callback. This is done so that the caller can reuse the same function as their callback without worrying about whether
or not it'll unsubscribe all watchers.

Setting changes are emitted into the default `WatchManager`, which calculates the new value for the setting. Ideally,
we'd also try and suppress updates which don't have a consequence on this value, however there's not an easy way to do
this. Instead, we just dispatch an update for all changes and leave it up to the consumer to deduplicate.

In practice, handlers which rely on remote changes (account data, room events, etc) will always attach a listener to the
`MatrixClient`. They then watch for changes to events they care about and send off appropriate updates to the
generalized `WatchManager` - a class specifically designed to deduplicate the logic of managing watchers. The handlers
which are localized to the local client (device) generally just trigger the `WatchManager` when they manipulate the
setting themselves as there's nothing to really 'watch'.
