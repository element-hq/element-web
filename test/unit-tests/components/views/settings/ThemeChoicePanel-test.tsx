/*
Copyright 2024 New Vector Ltd.
Copyright 2021 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import { act, render, screen, waitFor } from "jest-matrix-react";
import { mocked, MockedObject } from "jest-mock";
import userEvent from "@testing-library/user-event";
import fetchMock from "fetch-mock-jest";

import { ThemeChoicePanel } from "../../../../../src/components/views/settings/ThemeChoicePanel";
import SettingsStore from "../../../../../src/settings/SettingsStore";
import ThemeWatcher from "../../../../../src/settings/watchers/ThemeWatcher";
import { SettingLevel } from "../../../../../src/settings/SettingLevel";

jest.mock("../../../../../src/settings/watchers/ThemeWatcher");

describe("<ThemeChoicePanel />", () => {
    /**
     * Enable or disable the system theme
     * @param enable
     */
    async function enableSystemTheme(enable: boolean) {
        await SettingsStore.setValue("use_system_theme", null, SettingLevel.DEVICE, enable);
    }

    /**
     * Set the theme
     * @param theme
     */
    async function setTheme(theme: string) {
        await SettingsStore.setValue("theme", null, SettingLevel.DEVICE, theme);
    }

    beforeEach(async () => {
        mocked(ThemeWatcher).mockImplementation(() => {
            return {
                isSystemThemeSupported: jest.fn().mockReturnValue(true),
            } as unknown as MockedObject<ThemeWatcher>;
        });

        await enableSystemTheme(false);
        await setTheme("light");
    });

    it("renders the theme choice UI", () => {
        const { asFragment } = render(<ThemeChoicePanel />);
        expect(asFragment()).toMatchSnapshot();
    });

    describe("theme selection", () => {
        describe("system theme", () => {
            it("should disable Match system theme", async () => {
                render(<ThemeChoicePanel />);
                expect(screen.getByRole("checkbox", { name: "Match system theme" })).not.toBeChecked();
            });

            it("should enable Match system theme", async () => {
                await enableSystemTheme(true);

                render(<ThemeChoicePanel />);
                expect(screen.getByRole("checkbox", { name: "Match system theme" })).toBeChecked();
            });

            it("should change the system theme when clicked", async () => {
                jest.spyOn(SettingsStore, "setValue");

                render(<ThemeChoicePanel />);
                act(() => screen.getByRole("checkbox", { name: "Match system theme" }).click());

                // The system theme should be enabled
                expect(screen.getByRole("checkbox", { name: "Match system theme" })).toBeChecked();
                expect(SettingsStore.setValue).toHaveBeenCalledWith("use_system_theme", null, "device", true);
            });
        });

        describe("theme selection", () => {
            it("should disable theme selection when system theme is enabled", async () => {
                await enableSystemTheme(true);
                render(<ThemeChoicePanel />);

                // We expect all the themes to be disabled
                const themes = screen.getAllByRole("radio");
                themes.forEach((theme) => {
                    expect(theme).toBeDisabled();
                });
            });

            it("should enable theme selection when system theme is disabled", async () => {
                render(<ThemeChoicePanel />);

                // We expect all the themes to be disabled
                const themes = screen.getAllByRole("radio");
                themes.forEach((theme) => {
                    expect(theme).not.toBeDisabled();
                });
            });

            it("should have light theme selected", async () => {
                render(<ThemeChoicePanel />);

                // We expect the light theme to be selected
                const lightTheme = screen.getByRole("radio", { name: "Light" });
                expect(lightTheme).toBeChecked();

                // And the dark theme shouldn't be selected
                const darkTheme = screen.getByRole("radio", { name: "Dark" });
                expect(darkTheme).not.toBeChecked();
            });

            it("should switch to dark theme", async () => {
                jest.spyOn(SettingsStore, "setValue");

                render(<ThemeChoicePanel />);

                const darkTheme = screen.getByRole("radio", { name: "Dark" });
                const lightTheme = screen.getByRole("radio", { name: "Light" });
                expect(darkTheme).not.toBeChecked();

                // Switch to the dark theme
                act(() => darkTheme.click());
                expect(SettingsStore.setValue).toHaveBeenCalledWith("theme", null, "device", "dark");

                // Dark theme is now selected
                await waitFor(() => expect(darkTheme).toBeChecked());
                // Light theme is not selected anymore
                expect(lightTheme).not.toBeChecked();
                // The setting should be updated
                expect(SettingsStore.setValue).toHaveBeenCalledWith("theme", null, "device", "dark");
            });
        });
    });

    describe("custom theme", () => {
        const aliceTheme = { name: "Alice theme", is_dark: true, colors: {} };
        const bobTheme = { name: "Bob theme", is_dark: false, colors: {} };

        beforeEach(async () => {
            await SettingsStore.setValue("feature_custom_themes", null, SettingLevel.DEVICE, true);
            await SettingsStore.setValue("custom_themes", null, SettingLevel.DEVICE, [aliceTheme]);
        });

        it("should render the custom theme section", () => {
            const { asFragment } = render(<ThemeChoicePanel />);
            expect(asFragment()).toMatchSnapshot();
        });

        it("should add a custom theme", async () => {
            jest.spyOn(SettingsStore, "setValue");
            // Respond to the theme request
            fetchMock.get("http://bob.theme", {
                body: bobTheme,
            });

            render(<ThemeChoicePanel />);

            // Add the new custom theme
            const customThemeInput = screen.getByRole("textbox", { name: "Add custom theme" });
            await userEvent.type(customThemeInput, "http://bob.theme");
            screen.getByRole("button", { name: "Add custom theme" }).click();

            // The new custom theme is added to the user's themes
            await waitFor(() =>
                expect(SettingsStore.setValue).toHaveBeenCalledWith("custom_themes", null, "account", [
                    aliceTheme,
                    bobTheme,
                ]),
            );
        });

        it("should display custom theme", () => {
            const { asFragment } = render(<ThemeChoicePanel />);

            expect(screen.getByRole("radio", { name: aliceTheme.name })).toBeInTheDocument();
            expect(screen.getByRole("listitem", { name: aliceTheme.name })).toBeInTheDocument();
            expect(asFragment()).toMatchSnapshot();
        });
    });
});
