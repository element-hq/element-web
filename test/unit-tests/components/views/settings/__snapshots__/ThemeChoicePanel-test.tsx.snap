// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ThemeChoicePanel /> custom theme should display custom theme 1`] = `
<DocumentFragment>
  <div
    class="mx_SettingsSubsection mx_SettingsSubsection_newUi"
    data-testid="themePanel"
  >
    <div
      class="mx_SettingsSubsectionHeading"
    >
      <h3
        class="mx_Heading_h3 mx_SettingsSubsectionHeading_heading"
      >
        Theme
      </h3>
    </div>
    <div
      class="mx_SettingsSubsection_content mx_SettingsSubsection_content_newUi"
    >
      <form
        class="_root_dgy0u_24"
      >
        <div
          class="_inline-field_dgy0u_40"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_qnvru_18"
            >
              <input
                class="_input_qnvru_32"
                id="radix-:r28:"
                name="systemTheme"
                title=""
                type="checkbox"
              />
              <div
                class="_ui_qnvru_42"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67"
              for="radix-:r28:"
            >
              Match system theme
            </label>
          </div>
        </div>
      </form>
      <form
        class="_root_dgy0u_24 mx_ThemeChoicePanel_ThemeSelectors"
      >
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector mx_ThemeChoicePanel_themeSelector_enabled cpd-theme-light"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                checked=""
                class="_input_1vw5h_26"
                id="radix-:r29:"
                name="themeSelector"
                title=""
                type="radio"
                value="light"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r29:"
            >
              Light
            </label>
          </div>
        </div>
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector cpd-theme-dark"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                class="_input_1vw5h_26"
                id="radix-:r2a:"
                name="themeSelector"
                title=""
                type="radio"
                value="dark"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r2a:"
            >
              Dark
            </label>
          </div>
        </div>
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector cpd-theme-light"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                class="_input_1vw5h_26"
                id="radix-:r2b:"
                name="themeSelector"
                title=""
                type="radio"
                value="light-high-contrast"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r2b:"
            >
              High contrast
            </label>
          </div>
        </div>
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector cpd-theme-dark"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                class="_input_1vw5h_26"
                id="radix-:r2c:"
                name="themeSelector"
                title=""
                type="radio"
                value="custom-Alice theme"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r2c:"
            >
              Alice theme
            </label>
          </div>
        </div>
      </form>
      <div
        class="mx_ThemeChoicePanel_CustomTheme"
      >
        <form
          class="_root_dgy0u_24 mx_ThemeChoicePanel_CustomTheme_EditInPlace"
        >
          <div
            class="_field_dgy0u_34"
          >
            <label
              class="_label_dgy0u_67"
              for="radix-:r2d:"
            >
              Add custom theme
            </label>
            <div
              class="_controls_1h4nb_17"
            >
              <input
                aria-describedby="radix-:r2e:"
                class="_control_9gon8_18"
                id="radix-:r2d:"
                name="input"
                title=""
                value=""
              />
            </div>
            <span
              class="_message_dgy0u_98 _help-message_dgy0u_104"
              id="radix-:r2e:"
            >
              Enter the URL of a custom theme you want to apply.
            </span>
          </div>
        </form>
        <ul
          class="mx_ThemeChoicePanel_CustomThemeList"
        >
          <li
            aria-label="Alice theme"
            class="mx_ThemeChoicePanel_CustomThemeList_theme"
          >
            <span
              class="mx_ThemeChoicePanel_CustomThemeList_name"
            >
              Alice theme
            </span>
            <button
              aria-label="Delete"
              aria-labelledby=":r2f:"
              class="_icon-button_bh2qc_17 _destructive_bh2qc_83"
              role="button"
              style="--cpd-icon-button-size: 32px;"
              tabindex="0"
            >
              <div
                class="_indicator-icon_133tf_26"
                style="--cpd-icon-button-size: 100%;"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 21c-.55 0-1.02-.196-1.412-.587A1.926 1.926 0 0 1 5 19V6a.968.968 0 0 1-.713-.287A.968.968 0 0 1 4 5c0-.283.096-.52.287-.713A.968.968 0 0 1 5 4h4a.97.97 0 0 1 .287-.712A.968.968 0 0 1 10 3h4a.97.97 0 0 1 .713.288A.968.968 0 0 1 15 4h4a.97.97 0 0 1 .712.287c.192.192.288.43.288.713s-.096.52-.288.713A.968.968 0 0 1 19 6v13c0 .55-.196 1.02-.587 1.413A1.926 1.926 0 0 1 17 21H7ZM7 6v13h10V6H7Zm2 10c0 .283.096.52.287.712.192.192.43.288.713.288s.52-.096.713-.288A.968.968 0 0 0 11 16V9a.967.967 0 0 0-.287-.713A.968.968 0 0 0 10 8a.968.968 0 0 0-.713.287A.968.968 0 0 0 9 9v7Zm4 0c0 .283.096.52.287.712.192.192.43.288.713.288s.52-.096.713-.288A.968.968 0 0 0 15 16V9a.967.967 0 0 0-.287-.713A.968.968 0 0 0 14 8a.968.968 0 0 0-.713.287A.967.967 0 0 0 13 9v7Z"
                  />
                </svg>
              </div>
            </button>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="_separator_144s5_17"
      data-kind="primary"
      data-orientation="horizontal"
      role="separator"
    />
  </div>
</DocumentFragment>
`;

exports[`<ThemeChoicePanel /> custom theme should render the custom theme section 1`] = `
<DocumentFragment>
  <div
    class="mx_SettingsSubsection mx_SettingsSubsection_newUi"
    data-testid="themePanel"
  >
    <div
      class="mx_SettingsSubsectionHeading"
    >
      <h3
        class="mx_Heading_h3 mx_SettingsSubsectionHeading_heading"
      >
        Theme
      </h3>
    </div>
    <div
      class="mx_SettingsSubsection_content mx_SettingsSubsection_content_newUi"
    >
      <form
        class="_root_dgy0u_24"
      >
        <div
          class="_inline-field_dgy0u_40"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_qnvru_18"
            >
              <input
                class="_input_qnvru_32"
                id="radix-:r10:"
                name="systemTheme"
                title=""
                type="checkbox"
              />
              <div
                class="_ui_qnvru_42"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67"
              for="radix-:r10:"
            >
              Match system theme
            </label>
          </div>
        </div>
      </form>
      <form
        class="_root_dgy0u_24 mx_ThemeChoicePanel_ThemeSelectors"
      >
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector mx_ThemeChoicePanel_themeSelector_enabled cpd-theme-light"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                checked=""
                class="_input_1vw5h_26"
                id="radix-:r11:"
                name="themeSelector"
                title=""
                type="radio"
                value="light"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r11:"
            >
              Light
            </label>
          </div>
        </div>
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector cpd-theme-dark"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                class="_input_1vw5h_26"
                id="radix-:r12:"
                name="themeSelector"
                title=""
                type="radio"
                value="dark"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r12:"
            >
              Dark
            </label>
          </div>
        </div>
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector cpd-theme-light"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                class="_input_1vw5h_26"
                id="radix-:r13:"
                name="themeSelector"
                title=""
                type="radio"
                value="light-high-contrast"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r13:"
            >
              High contrast
            </label>
          </div>
        </div>
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector cpd-theme-dark"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                class="_input_1vw5h_26"
                id="radix-:r14:"
                name="themeSelector"
                title=""
                type="radio"
                value="custom-Alice theme"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r14:"
            >
              Alice theme
            </label>
          </div>
        </div>
      </form>
      <div
        class="mx_ThemeChoicePanel_CustomTheme"
      >
        <form
          class="_root_dgy0u_24 mx_ThemeChoicePanel_CustomTheme_EditInPlace"
        >
          <div
            class="_field_dgy0u_34"
          >
            <label
              class="_label_dgy0u_67"
              for="radix-:r15:"
            >
              Add custom theme
            </label>
            <div
              class="_controls_1h4nb_17"
            >
              <input
                aria-describedby="radix-:r16:"
                class="_control_9gon8_18"
                id="radix-:r15:"
                name="input"
                title=""
                value=""
              />
            </div>
            <span
              class="_message_dgy0u_98 _help-message_dgy0u_104"
              id="radix-:r16:"
            >
              Enter the URL of a custom theme you want to apply.
            </span>
          </div>
        </form>
        <ul
          class="mx_ThemeChoicePanel_CustomThemeList"
        >
          <li
            aria-label="Alice theme"
            class="mx_ThemeChoicePanel_CustomThemeList_theme"
          >
            <span
              class="mx_ThemeChoicePanel_CustomThemeList_name"
            >
              Alice theme
            </span>
            <button
              aria-label="Delete"
              aria-labelledby=":r17:"
              class="_icon-button_bh2qc_17 _destructive_bh2qc_83"
              role="button"
              style="--cpd-icon-button-size: 32px;"
              tabindex="0"
            >
              <div
                class="_indicator-icon_133tf_26"
                style="--cpd-icon-button-size: 100%;"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  viewBox="0 0 24 24"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M7 21c-.55 0-1.02-.196-1.412-.587A1.926 1.926 0 0 1 5 19V6a.968.968 0 0 1-.713-.287A.968.968 0 0 1 4 5c0-.283.096-.52.287-.713A.968.968 0 0 1 5 4h4a.97.97 0 0 1 .287-.712A.968.968 0 0 1 10 3h4a.97.97 0 0 1 .713.288A.968.968 0 0 1 15 4h4a.97.97 0 0 1 .712.287c.192.192.288.43.288.713s-.096.52-.288.713A.968.968 0 0 1 19 6v13c0 .55-.196 1.02-.587 1.413A1.926 1.926 0 0 1 17 21H7ZM7 6v13h10V6H7Zm2 10c0 .283.096.52.287.712.192.192.43.288.713.288s.52-.096.713-.288A.968.968 0 0 0 11 16V9a.967.967 0 0 0-.287-.713A.968.968 0 0 0 10 8a.968.968 0 0 0-.713.287A.968.968 0 0 0 9 9v7Zm4 0c0 .283.096.52.287.712.192.192.43.288.713.288s.52-.096.713-.288A.968.968 0 0 0 15 16V9a.967.967 0 0 0-.287-.713A.968.968 0 0 0 14 8a.968.968 0 0 0-.713.287A.967.967 0 0 0 13 9v7Z"
                  />
                </svg>
              </div>
            </button>
          </li>
        </ul>
      </div>
    </div>
    <div
      class="_separator_144s5_17"
      data-kind="primary"
      data-orientation="horizontal"
      role="separator"
    />
  </div>
</DocumentFragment>
`;

exports[`<ThemeChoicePanel /> renders the theme choice UI 1`] = `
<DocumentFragment>
  <div
    class="mx_SettingsSubsection mx_SettingsSubsection_newUi"
    data-testid="themePanel"
  >
    <div
      class="mx_SettingsSubsectionHeading"
    >
      <h3
        class="mx_Heading_h3 mx_SettingsSubsectionHeading_heading"
      >
        Theme
      </h3>
    </div>
    <div
      class="mx_SettingsSubsection_content mx_SettingsSubsection_content_newUi"
    >
      <form
        class="_root_dgy0u_24"
      >
        <div
          class="_inline-field_dgy0u_40"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_qnvru_18"
            >
              <input
                class="_input_qnvru_32"
                id="radix-:r0:"
                name="systemTheme"
                title=""
                type="checkbox"
              />
              <div
                class="_ui_qnvru_42"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67"
              for="radix-:r0:"
            >
              Match system theme
            </label>
          </div>
        </div>
      </form>
      <form
        class="_root_dgy0u_24 mx_ThemeChoicePanel_ThemeSelectors"
      >
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector mx_ThemeChoicePanel_themeSelector_enabled cpd-theme-light"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                checked=""
                class="_input_1vw5h_26"
                id="radix-:r1:"
                name="themeSelector"
                title=""
                type="radio"
                value="light"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r1:"
            >
              Light
            </label>
          </div>
        </div>
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector cpd-theme-dark"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                class="_input_1vw5h_26"
                id="radix-:r2:"
                name="themeSelector"
                title=""
                type="radio"
                value="dark"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r2:"
            >
              Dark
            </label>
          </div>
        </div>
        <div
          class="_inline-field_dgy0u_40 mx_ThemeChoicePanel_themeSelector cpd-theme-light"
        >
          <div
            class="_inline-field-control_dgy0u_52"
          >
            <div
              class="_container_1vw5h_18"
            >
              <input
                class="_input_1vw5h_26"
                id="radix-:r3:"
                name="themeSelector"
                title=""
                type="radio"
                value="light-high-contrast"
              />
              <div
                class="_ui_1vw5h_27"
              />
            </div>
          </div>
          <div
            class="_inline-field-body_dgy0u_46"
          >
            <label
              class="_label_dgy0u_67 mx_ThemeChoicePanel_themeSelector_Label"
              for="radix-:r3:"
            >
              High contrast
            </label>
          </div>
        </div>
      </form>
    </div>
    <div
      class="_separator_144s5_17"
      data-kind="primary"
      data-orientation="horizontal"
      role="separator"
    />
  </div>
</DocumentFragment>
`;
